ipn/ipnlocal: data race during TestAutoExitNodeSetNetInfoCallback

At 2506ede862 (which was green before merging) ... 


https://github.com/tailscale/tailscale/actions/runs/29287115604/job/86942231045
```
=== RUN   TestAutoExitNodeSetNetInfoCallback
    local_test.go:477: Added memory store for testing
    local_test.go:486: Added fake userspace engine for testing
==================
WARNING: DATA RACE
Read at 0x00c0003424e8 by goroutine 2955:
  tailscale.com/net/netcheck.(*Client).timeNow()
      /home/runner/work/tailscale/tailscale/src/net/netcheck/netcheck.go:1325 +0x39
  tailscale.com/net/netcheck.(*Client).GetReport()
      /home/runner/work/tailscale/tailscale/src/net/netcheck/netcheck.go:834 +0x34c
  tailscale.com/wgengine/magicsock.(*Conn).updateNetInfo()
      /home/runner/work/tailscale/tailscale/src/wgengine/magicsock/magicsock.go:1036 +0x2b8
  tailscale.com/wgengine/magicsock.(*Conn).determineEndpoints()
      /home/runner/work/tailscale/tailscale/src/wgengine/magicsock/magicsock.go:1293 +0x129
  tailscale.com/wgengine/magicsock.(*Conn).updateEndpoints()
      /home/runner/work/tailscale/tailscale/src/wgengine/magicsock/magicsock.go:939 +0x2a4
  tailscale.com/wgengine/magicsock.(*Conn).ReSTUN.gowrap2()
      /home/runner/work/tailscale/tailscale/src/wgengine/magicsock/magicsock.go:3684 +0x46

Previous write at 0x00c0003424e8 by goroutine 2913:
  tailscale.com/net/netcheck.(*Client).AddReportHistoryForTest.func1()
      /home/runner/work/tailscale/tailscale/src/net/netcheck/netcheck.go:1520 +0x35
  tailscale.com/net/netcheck.(*Client).AddReportHistoryForTest.deferwrap1()
      /home/runner/work/tailscale/tailscale/src/net/netcheck/netcheck.go:1520 +0x38
  runtime.deferreturn()
      /home/runner/.cache/tsgo/63ae404c8203317fd3c82d972e5dc8f0fcb425cb/src/runtime/panic.go:668 +0x5d
  tailscale.com/wgengine/magicsock.(*Conn).AddNetcheckReportForTest()
      /home/runner/work/tailscale/tailscale/src/wgengine/magicsock/magicsock.go:4353 +0xa5
  tailscale.com/ipn/ipnlocal.TestAutoExitNodeSetNetInfoCallback()
      /home/runner/work/tailscale/tailscale/src/ipn/ipnlocal/local_test.go:4232 +0x18a5
  testing.tRunner()
      /home/runner/.cache/tsgo/63ae404c8203317fd3c82d972e5dc8f0fcb425cb/src/testing/testing.go:2036 +0x21c
  testing.(*T).Run.gowrap1()
      /home/runner/.cache/tsgo/63ae404c8203317fd3c82d972e5dc8f0fcb425cb/src/testing/testing.go:2101 +0x38

Goroutine 2955 (running) created at:
  tailscale.com/wgengine/magicsock.(*Conn).ReSTUN()
      /home/runner/work/tailscale/tailscale/src/wgengine/magicsock/magicsock.go:3684 +0x5bc
  tailscale.com/wgengine/magicsock.(*Conn).setDERPMap.gowrap2()
      /home/runner/work/tailscale/tailscale/src/wgengine/magicsock/derp.go:873 +0x3a

Goroutine 2913 (running) created at:
  testing.(*T).Run()
      /home/runner/.cache/tsgo/63ae404c8203317fd3c82d972e5dc8f0fcb425cb/src/testing/testing.go:2101 +0xb12
  testing.runTests.func1()
      /home/runner/.cache/tsgo/63ae404c8203317fd3c82d972e5dc8f0fcb425cb/src/testing/testing.go:2585 +0x84
  testing.tRunner()
      /home/runner/.cache/tsgo/63ae404c8203317fd3c82d972e5dc8f0fcb425cb/src/testing/testing.go:2036 +0x21c
  testing.runTests()
      /home/runner/.cache/tsgo/63ae404c8203317fd3c82d972e5dc8f0fcb425cb/src/testing/testing.go:2583 +0x9e9
  testing.(*M).Run()
      /home/runner/.cache/tsgo/63ae404c8203317fd3c82d972e5dc8f0fcb425cb/src/testing/testing.go:2443 +0xf4b
  main.main()
      _testmain.go:414 +0x164
==================
    testing.go:1712: race detected during execution of test
--- FAIL: TestAutoExitNodeSetNetInfoCallback (0.27s)
ok	tailscale.com/ipn/ipnstate	(cached)

```

--- a/net/netcheck/netcheck.go
+++ b/net/netcheck/netcheck.go
@@ -1026,7 +1026,7 @@ func (c *Client) finishAndStoreReport(rs *reportState, dm *tailcfg.DERPMap) *Rep
 	report := rs.report.Clone()
 	rs.mu.Unlock()
 
-	c.addReportHistoryAndSetPreferredDERP(rs, report, dm.View())
+	c.addReportHistoryAndSetPreferredDERP(rs, report, dm.View(), c.timeNow())
 	c.logConciseReport(report, dm)
 
 	return report
@@ -1381,7 +1381,7 @@ func (c *Client) addReportAndPruneExpired(now time.Time, r *Report) {
 
 // addReportHistoryAndSetPreferredDERP adds r to the set of recent Reports
 // and mutates r.PreferredDERP to contain the best recent one.
-func (c *Client) addReportHistoryAndSetPreferredDERP(rs *reportState, r *Report, dm tailcfg.DERPMapView) {
+func (c *Client) addReportHistoryAndSetPreferredDERP(rs *reportState, r *Report, dm tailcfg.DERPMapView, now time.Time) {
 	c.mu.Lock()
 	defer c.mu.Unlock()
 
@@ -1392,7 +1392,6 @@ func (c *Client) addReportHistoryAndSetPreferredDERP(rs *reportState, r *Report,
 
 	// Add report to history, enforce retention window, then take the best (lowest)
 	// latency seen per region across what remains.
-	now := c.timeNow()
 	c.addReportAndPruneExpired(now, r)
 	bestRecent := c.bestRecentLatencyLocked()
 
@@ -1517,10 +1516,8 @@ func (c *Client) RecentRegionLatency() map[int]time.Duration {
 // r.PreferredDERP from that history.
 func (c *Client) AddReportHistoryForTest(dm *tailcfg.DERPMap, r *Report, now time.Time) {
 	testenv.AssertInTest()
-	defer func(prev func() time.Time) { c.TimeNow = prev }(c.TimeNow)
-	c.TimeNow = func() time.Time { return now }
 	rs := &reportState{c: c, start: now}
-	c.addReportHistoryAndSetPreferredDERP(rs, r, dm.View())
+	c.addReportHistoryAndSetPreferredDERP(rs, r, dm.View(), now)
 }
 
 func updateLatency(m map[int]time.Duration, regionID int, d time.Duration) {

--- a/net/netcheck/netcheck_test.go
+++ b/net/netcheck/netcheck_test.go
@@ -468,7 +468,7 @@ func TestAddReportHistoryAndSetPreferredDERP(t *testing.T) {
 			for _, s := range tt.steps {
 				fakeTime = fakeTime.Add(s.after)
 				rs.start = fakeTime.Add(-100 * time.Millisecond)
-				c.addReportHistoryAndSetPreferredDERP(rs, s.r, dm.View())
+				c.addReportHistoryAndSetPreferredDERP(rs, s.r, dm.View(), fakeTime)
 			}
 			lastReport := tt.steps[len(tt.steps)-1].r
 			if got, want := len(c.prev), tt.wantPrevLen; got != want {
@@ -522,7 +522,7 @@ func TestRecentReportsRetainFullNetcheck(t *testing.T) {
 			regions = allRegions
 			lastFull = now
 		}
-		c.addReportHistoryAndSetPreferredDERP(&reportState{c: c, start: now}, mkReport(regions), dm.View())
+		c.addReportHistoryAndSetPreferredDERP(&reportState{c: c, start: now}, mkReport(regions), dm.View(), now)
 
 		// Recent latency must always cover every region, which is only
 		// possible while a full report remains in c.prev.

